File: src/main/java/com/oycm/algorithm/h/basic/Solution_9.java
package com.oycm.algorithm.h.basic;


public class Solution_9 {

    /**
     * 868. <a href="https://leetcode.cn/problems/binary-gap/description/">二进制间距</a> 1307
     * <p>
     * 相邻: 只有 0 将两个 1 分隔开（可能不存在 0 ）
     *
     * @param n
     * @return 求 n 的二进制表示中两个 相邻 1 之间的 最长距离
     */
    public int binaryGap(int n) {
        /*
        二进制中 附近 1 之间的最长距离
         */
        if (Integer.bitCount(n) == 1) {
            return 0;
        }
        int ans = 0;
        int distance = 0;
        while (n > 0) {
            if ((n & 1) == 1) {
                distance = 1;
            } else if (distance > 0) {
                distance++;
            }
            n = n >> 1;
            ans = Math.max(ans, distance);
        }
        return ans;
    }
}

File: src/main/java/com/oycm/algorithm/e/enum_right_log_left/basic/Solution_9.scala
package com.oycm.algorithm.e.enum_right_log_left.basic

object Solution_9 {

  /**
   * 2260. 必须拿起的最小连续卡牌数 1365
   * https://leetcode.cn/problems/minimum-consecutive-cards-to-pick-up/description/
   *
   * cards[i] 表示第 i 张卡牌的 值。
   * 如果两张卡牌的值相同，则认为这一对卡牌 匹配 。
   *
   * @param cards 整数数组
   * @return 求拿起的最小连续卡牌数，以使在拿起的卡牌中有一对匹配的卡牌。否则返回 -1
   */
  def minimumCardPickup(cards: Array[Int]): Int = {
    /*
    遍历过程中，用数组维护 值的下标，如果前面存在该值的卡牌，计算答案，更新下标，因为越和后面的成对值越大
    */
    var ans = Int.MaxValue
    val map = scala.collection.mutable.Map[Int, Int]()
    for (i <- cards.indices if ans > 2) {
      val card = cards(i)
      if (map.contains(card)) {
        ans = Math.min(ans, i - map(card) + 1)
      }
      map(card) = i
    }
    if (ans == Int.MaxValue) {
      ans = -1
    }
    ans
  }

  def main(args: Array[String]): Unit = {
    println(minimumCardPickup(Array(3, 4, 2, 3, 4, 7)))
    println(minimumCardPickup(Array(1, 0, 5, 3)))
  }
}
